### PR TITLE
[犀牛鸟-E3]：五族混合系统的路由透视镜

### PR DESCRIPTION
**Pull request by Ricky-7-Yan**
## Summary

This PR adds the E3 routing-observability workflow under `tools/e3_routing` without changing model forward
signatures or routing decisions.

- **P0:** one versioned event schema and static diagnostics for real MoE, MoT and Latent forwards.
- **P1:** a read-only localhost dashboard plus paired full-step overhead measurement.
- **P2:** a five-family feasibility audit and original-image overlays only for routers with a genuine 2D token axis.

Hooks are removable and hooked/unhooked outputs are checked for equality. Spatial diagnostics fail closed: MoE,
Latent and MoLoRA are reported as `UNSUPPORTED` instead of reshaping sample-level or expert-axis vectors into heatmaps.

## Acceptance results

| Level | Result |
| --- | --- |
| P0 | 52 primary events across 6 MoE, 4 MoT and 3 Latent modules; 52/52 repeat fingerprints matched; batch 2/4 maximum load delta `0` (`1e-5` gate) |
| P1 | Three-family HTTP dashboard; 108 paired full training steps; all slowdown medians and 95% CI upper bounds below the predeclared 10% gate |
| P2 | Five families audited; MoT/MoA spatially supported; 32 captures, 240 raw arrays and 192 selectable views; hook output delta `0` |

### P1 overhead

| Family | Pairs | Median slowdown | 95% bootstrap CI | Verdict |
| --- | ---: | ---: | ---: | --- |
| MoE | 30 | 2.113% | [-2.514%, 8.361%] | PASS / STRONG |
| MoT | 48 | -0.050% | [-6.374%, 9.451%] | PASS / STRONG |
| Latent | 30 | -5.741% | [-16.195%, 1.570%] | PASS / STRONG |

Negative values are treated as CPU timing noise, not observer speedups. The measured step includes detection loss,
backward, gradient clipping, grouped SGD, and the requested JSONL write/flush.

### P0

- Defines `e3.routing/v1.0.0` with expert load, entropy, Gini, dominant share, mixing-weight availability and
  auxiliary-loss availability.
- Writes JSONL, aggregates, repeat/batch comparisons, static plots, resolved configuration, input hashes and
  SHA-256 manifests.

### P1

- Reads the same event stream through health, dashboard and snapshot endpoints bound to `127.0.0.1`.
- Separates source count, active window size and truncation; only a concurrently written final partial line is ignored.
- Compares `off`, `capture_memory` and `capture_jsonl` under a balanced schedule and includes an official
  `DetectionTrainer` state-equivalence cross-check.

### P2

- Captures router-child outputs before public snapshots average spatial axes.
- Reverses recorded letterbox geometry and exports fixed-scale probability, entropy, Top-1 margin and
  dominant-expert views at the original image size.
- Adds batch-position identity, repeatability, geometry, source-fingerprint, ground-truth-region and asset-integrity
  gates.
- Includes a localhost demo with family/sample/layer/view selection, ground-truth toggling, metadata copy and PNG
  export.

The extended diagnostic study covers 32 images, three seeds and graded brightness/contrast/blur perturbations. The
predeclared blur residual test produced no Holm-significant association; the null result is retained rather than
reinterpreted after the fact.

## Verification

Final release tree: `c753ca3d13ab6f90bba86636774ce7e4c937c47d`, based directly on upstream
`7bfbfd374a6b44720f98366e088f3a962e16321d`.

- `113 passed` for the combined phase and CLI-contract suite.
- Ruff: `PASS` for `src`, `tests` and `scripts`.
- Upstream config/model regression subset: `13 passed`; one expected MoA head-adjustment warning.
- Latest-main real-forward smoke: `PASS`; P0 module counts `6/4/3`, MoT/MoA spatial counts `4/4`, all hook deltas `0`.
- Latest-main full P0: `PASS`; 52 primary + 52 repeat events, batch 2/4 maximum load delta `0`.
- P1 dashboard over the latest P0 stream: HTTP 200, 52 events, MoE/MoT/Latent, no truncation.
- Latest-main full P2 with `require_committed_source=true`: `PASS`; clean implementation state and 9/9 source
  fingerprints matched.
- Seven archived formal evidence packages: 395 manifest entries independently rehashed with zero mismatch.

## Reviewer quick start

From the YOLO-Master repository root:

```bat
cd tools\e3_routing
python -m pip install -e ".[test]"
python -m pytest tests -q
python -m ruff check src tests scripts
python scripts\latest_main_smoke.py --repo-root ..\.. --output artifacts\latest-main-smoke.json
python -m e3_p0 --run-id review-p0 --no-update-latest
python -m e3_p2 run --run-id review-p2 --no-latest
python -m e3_p1.dashboard --events artifacts\p0\review-p0\routing-all.jsonl
```

Active configs target the containing checkout. Frozen configs preserve the source identity used for the archived
numerical evidence; latest-main compatibility checks do not relabel historical runs.

## Evidence

- P0 fixed evidence: https://github.com/Ricky-7-Yan/YOLO-Master-E3-P0/tree/7766f5038d8cafb40f001dd0ce3fd9c9d89030b8
- P1 fixed evidence: https://github.com/Ricky-7-Yan/YOLO-Master-E3-P1/tree/bea6e7e526ca7667c1b933e8ac24627ce757d337
- P2 fixed evidence: https://github.com/Ricky-7-Yan/YOLO-Master-E3-P2/tree/b04c52286502cd41a437081c046166a9db091c7c

## Demo video

https://github.com/user-attachments/assets/9dbbcf77-5b89-40bb-ac48-25d412b68dba

## Limitations

- Evidence is CPU-only and primarily uses randomly initialized models. It validates instrumentation and bounded
  cold-start diagnostics, not trained AP, convergence or expert semantics.
- No CUDA overhead or GPU-memory claim is made.
- The short coco8 `DetectionTrainer` epoch windows are descriptive; the preregistered 108-pair full-step benchmark
  remains the formal P1 verdict.
- Only MoT and MoA expose recoverable 2D token routes in the audited runtime.
- Routing/output associations are diagnostic, not causal.

### CODE REVIEW

**Comment by Ricky-7-Yan** [comment]
CI triage for current head 75c939d: all four OS jobs reach the full upstream suite and fail on the same two tests outside 	ools/e3_routing — 	est_enabled_without_teacher_is_rejected (expected teacher error, receives cache-dir error) and 	est_response_kd_builds_pseudo_batch_from_cached_responses (TAL index 1 out of bounds). I compared the latest main push run at f961b99; its Ubuntu job reproduces both exact failures. This is therefore a current base-branch failure, not introduced by the E3-only diff. The PR-specific 113-test suite and Ruff checks remain green as recorded above.

**Comment by isLinXu** [comment]
P0/P1/P2 三级工具链，52 个路由事件覆盖 13 个模块，工作量没话说。PR body 的开销 bootstrap CI 区间和已知局限分析写得很详细。

体积有点大（14K 行 / 100 文件），`frozen/` 下面的实验配置我帮你整理一下。你确认一件事：52 个路由事件的测试跑一遍要多久？超过 5 分钟的话我标个 `--slow`。

这个会作为 E3 的主 PR 合入。

**Comment by Ricky-7-Yan** [comment]
导师，我刚刚复测了P0 流程，依赖和 coco8 数据已就绪。连续运行 3 次分别耗时 16.63 秒、15.34 秒和 15.31 秒。每次均生成 52 个 primary events 和 52 个 repeat events，52/52 重复一致，batch 2/4 consistency 也全部通过。 不需要标记slow

**Comment by isLinXu** [comment]
LGTM
